Source/ACE/Entity/Key.cs

```diff
@@ -36,26 +39,49 @@ public void HandleActionUseOnTarget(Player player, WorldObject target)
                     player.DoMoveTo(target);
                 else
                 {
-                    if (Structure == 0)
-                    {
-                        var sendUseDoneEvent = new GameEventUseDone(player.Session);
-                        player.Session.Network.EnqueueSend(sendUseDoneEvent);
-                    }
-                    else
+                    var sendUseDoneEvent = new GameEventUseDone(player.Session);
+                    if (target.WeenieType == WeenieType.Door)
                     {
                         Door door = target as Door;
-                        var sendUseDoneEvent = new GameEventUseDone(player.Session);
-                        if (door.UnlockDoor(KeyCode))
-                        {
-                            Structure--;
-                            player.Session.Network.EnqueueSend(sendUseDoneEvent);
-                        }
-                        else
+                        Door.UnlockDoorResults results = door.UnlockDoor(KeyCode);
+                        switch (results)
                         {
-                            var message = new GameMessageSystemChat($"The {Name} cannot be used on the {target.Name}.", ChatMessageType.System);
-                            player.Session.Network.EnqueueSend(sendUseDoneEvent, message);
+                            case Entity.Door.UnlockDoorResults.UnlockSuccess:
+                                Structure--;
+                                SendPartialUpdates(player.Session, _updateStructure);
+                                if (Structure < 1)
+                                {
+                                    // Remove item from players
+                                    player.DestroyInventoryItem(this);
+                                    // Clean up the shard database.
+                                    DatabaseManager.Shard.DeleteObject(SnapShotOfAceObject(), null);
+                                }
+                                player.Session.Network.EnqueueSend(sendUseDoneEvent);
+                                break;
+                            case Entity.Door.UnlockDoorResults.DoorOpen:
+                                var messageDoorOpen = new GameEventDisplayStatusMessage(player.Session, StatusMessageType1.Enum_0481);
```

Comment:
0481?  If you can come up with a better name for that, please do.

Reply:
That is just the name of the Enum from StatusMessageType1.cs already present.  I did not create that.  Technically the message associated with that enum isn't even correct, anyways.  The client displayed "The monarch doesn't own either a mansion or villia", instead of "You cannot unlock or lock open doors" as defined in StatusMessageType1.cs